### ISSUE TITLE
SQL Server to SIARDDK: Process stops after views and does not encorporate views

### ISSUE THREAD

**Issue by PhillipTommerholt** [keeps/db-preservation-toolkit#194]
Description: 
When I try to export Northwind database on MS SQL Server 2014 to SIARDDK, the export stops somewhere soon after export of the last view in the database according to the two log files:

[dbptk-app.log.txt](https://github.com/keeps/db-preservation-toolkit/files/260756/dbptk-app.log.txt)
[dbptk-report-20160512084025398.txt](https://github.com/keeps/db-preservation-toolkit/files/260749/dbptk-report-20160512084025398.txt)

When I look at the produced tableIndex.xml I can see that the last table is missing the the ending elements, so there might be problems with handling where tables end and views begin.
[tableIndex.zip](https://github.com/keeps/db-preservation-toolkit/files/260729/tableIndex.zip)

**Comment by PhillipTommerholt** [comment]
@andreaskring

**Comment by andreaskring** [comment]
@PhillipTommerholt I think the problem lies within the microsoft-sql-server import module. When exporting the Northwind database from MySQL everything works fine. Previously, a problem with views also occured for MySQL, since these were not extracted from the DB by the MySQL import module. I think this problem may be similar. @chalkos can you take a look at this?

**Comment by chalkos** [comment]
The log ends with error which root cause is in the log as:

```
Caused by: org.xml.sax.SAXParseException: cvc-complex-type.2.4.b:
The content of element 'view' is not complete.
One of '{"http://www.sa.dk/xmlns/diark/1.0":queryOriginal}' is expected.
```

It seems that SIARDDK XML file should have a `queryOriginal` element inside every `view` element and that is not happening (maybe its null).

I have successfully imported views in the MS SQL module and exported them in SIARD1 and 2.
@andreaskring Please confirm that the fault is not in SIARDDK. If you can trace the problem back to MS SQL module, tell me and I will investigate this further.

**Comment by andreaskring** [comment]
@chalkos @PhillipTommerholt I had another look at this, and as far as I can see the problem is not within the SIARDDK code. As you describe above, the marshalling fails since the `queryOriginal` is missing.

The `queryOriginal` is handled [here](https://github.com/keeps/db-preservation-toolkit/blob/master/dbptk-core/src/main/java/com/databasepreservation/modules/siard/out/metadata/TableIndexFileStrategy.java#L245). The `queryOriginal` is mandatory in SIARDDK and it is fetched from the [ViewStructure](https://github.com/keeps/db-preservation-toolkit/blob/master/dbptk-model/src/main/java/com/databasepreservation/model/structure/ViewStructure.java) object by calling the method `getQueryOriginal()`, but the `queryOriginal` is never set in the [JDBCImportModule](https://github.com/keeps/db-preservation-toolkit/blob/master/dbptk-core/src/main/java/com/databasepreservation/modules/jdbc/in/JDBCImportModule.java) or in the [SQLServerJDBCImportModule](https://github.com/keeps/db-preservation-toolkit/blob/master/dbptk-core/src/main/java/com/databasepreservation/modules/sqlServer/in/SQLServerJDBCImportModule.java), so I guess this is why the marshalling fails. As mentioned previously, there are no problem when exporting the same database (Northwind) from a MySQL DBMS (the `queryOriginal` is set here for the [MySQLJDBCImportModule](https://github.com/keeps/db-preservation-toolkit/blob/master/dbptk-core/src/main/java/com/databasepreservation/modules/mySql/in/MySQLJDBCImportModule.java#L199).

@chalkos Could you have another look at this - do you agree that the problem occurs due to the missing call to the `setQueryOriginal()` method?

**Comment by chalkos** [comment]
Yes. I will fix this.

**Comment by andreaskring** [comment]
@chalkos @PhillipTommerholt Great - thank you.

**Comment by chalkos** [comment]
The fix I implemented tries to obtain the SQL and if it can't it sets the `originalQuery` to "unknown" (may be changed if you believe there is a better value).

Limitations of MSSQL server may get in the way of getting valid SQL statements:
(source http://stackoverflow.com/a/4765478)

> Caveats
> 
> * If the view was last modified with ALTER VIEW, then the script will be an ALTER VIEW statement rather than a CREATE VIEW statement.
> * The script reflects the name as it was created. The only time it gets updated is if you execute ALTER VIEW, or drop and recreate the view with CREATE VIEW. If the view has been renamed (e.g., via sp_rename) or ownership has been transferred to a different schema, the script you get back will reflect the original CREATE/ALTER VIEW statement: it will not reflect the objects current name.
> * Some tools truncate the output. For example, the MS-SQL command line tool sqlcmd.exe truncates the data at 255 chars. You can pass the parameter -y N to get the result with N chars.

These are the main reasons why I added a default value for `originalQuery`.
@PhillipTommerholt is "unknown" a good value to use instead of the SQL query when DBPTK can not obtain the original SQL statement? The report informs that the original SQL statement could not be obtained.

**Comment by PhillipTommerholt** [comment]
Hi,
I never answered this one. Just to be safe:

DBPTK tries to retrieve the original query no matter if it is valid or not, and:

**If it can:** 
In SIARD2.0 it gives the result in metadata.xml in the specific view with the element: `<queryOriginal>select * bla bla bla</queryOriginal>`
In SIARDK DK it gives the result in tableIndex.xml in the specific view with the element: `<queryOriginal>select * bla bla bla</queryOriginal>`
**If it can't:** 
In SIARD2.0 it leaves it out since its optional
In SIARDK DK it gives the result in tableIndex.xml in the specific view with the element:  `<queryOriginal>unknown</queryOriginal>
`

?

**Comment by chalkos** [comment]
Then it is not working correctly, since queryOriginal is being saved as `unknown` in SIARD2.

Tasklist @chalkos:
- do not set `unknown` as default value

Tasklist @andreaskring:
- set `unknown` as default value when writing the XML if getQueryOriginal is blank.

**Comment by andreaskring** [comment]
The SIARDDK part of the issue has been fixed in [this](https://github.com/magenta-aps/db-preservation-toolkit/tree/issue194_213_241_242) branch. A jar containing the fix can be found [here](https://drive.google.com/open?id=0B8hAEqjI-ScYTmlBMm5WM1ZjMFU)

**Comment by PhillipTommerholt** [comment]
I have tested it up against Northwind. 
The SIARDDK now gives the default value `"unkown"` and the export succeeds. So in that sense the issue is fixed.
But I am unsure why DBPTK can't retrieve the original query. Just by simply looking in MSMS I can easily see the query for the first view:
![image](https://cloud.githubusercontent.com/assets/15984159/17433059/d32aea2c-5b02-11e6-968b-f15594d972c8.png)

@chalkos Just to be safe: I am not saying the SIARD1.0/2.0 export module can't have `"unknown" `as default. I am only saying that it is optional in the specifications, so it is up to the export tool itself if it leaves it out completely or give a default value.

**Comment by chalkos** [comment]
> @chalkos Just to be safe: I am not saying the SIARD1.0/2.0 export module can't have "unknown"as default. I am only saying that it is optional in the specifications, so it is up to the export tool itself if it leaves it out completely or give a default value.

Don't worry, I made the decision to remove the "unknown" for SIARD2 because it is optional. Not having a value is, in my opinion, better than having a "*ups, there was a problem*" value. Makes it clearer that something is missing.

**Comment by chalkos** [comment]
@andreaskring thanks. I'll fix my part after merging your changes into master.

**Comment by PhillipTommerholt** [comment]
@chalkos 
Have you seen this in my comment above:

> But I am unsure why DBPTK can't retrieve the original query. Just by simply looking in MSMS I can easily see the query for the first view:
https://cloud.githubusercontent.com/assets/15984159/17433059/d32aea2c-5b02-11e6-968b-f15594d972c8.png 

Should I open a new issue for this specific matter? It goes for both SIARD1.0, 2.0 and SIARDDK.

**Comment by chalkos** [comment]
@PhillipTommerholt I think that would be better. this issue was fixed and closed, then after all it needed some adjustments, so it was open again, then waiting on a pull request, which is now created... if we add more problems this will get too confusing very soon 😉
